File: 538-convert-bst-to-greater-tree/538-convert-bst-to-greater-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode convertBST(TreeNode root) {
        List<TreeNode> list = new ArrayList<>();
        inorder(root, list);
        int prev = 0;
        for(int i = list.size()  -1; i >= 0; i--) {
            TreeNode node = list.get(i);
            node.val = node.val + prev;
            prev = node.val;
        }
        return root;
    }
    public void inorder(TreeNode root, List<TreeNode> list) {
        if(root == null) return;
        inorder(root.left, list);
        list.add(root);
        inorder(root.right, list);
    }
}

File: 538-convert-bst-to-greater-tree/538-convert-bst-to-greater-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    #正常的中序遍历顺序是先左子树后右子树，这里需要反过来，先右子树后左子树
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        sum = 0
        def inorder(root):
            nonlocal sum
            if not root:
                return
            inorder(root.right)
            sum += root.val
            root.val = sum
            inorder(root.left)
            
        inorder(root)
        return root